openmc/kinetics/solver.py

```diff
@@ -0,0 +1,869 @@
+            'ADIABATIC' approximates the time-derivatives as 0.
+
+        """
+
+        self._setup_openmc(time_point, method)
+
+        # Names of the statepoint and summary files
+        sp_old_name = self.directory / 'statepoint.{}.h5'.format(self.settings.batches)
+        sp_new_name = self.directory / 'statepoint_{:.6f}_sec.{}.h5'\
+                      .format(self.clock.times[time_point], self.settings.batches)
+        sum_old_name = self.directory / 'summary.h5'
+        sum_new_name = self.directory / 'summary_{:.6f}_sec.h5'.format(self.clock.times[time_point])
+        old_source = self.directory / 'source.{}.h5'.format(self.settings.batches)
+        new_source = self.directory / 'source.h5'
+
+        # Run OpenMC
+        if not self.use_pregenerated_sps:
+            openmc.run(cwd=self.directory, **self.run_kwargs)
+
+            # Rename the statepoint and summary files
+            copyfile(sp_old_name, sp_new_name)
+            copyfile(sum_old_name, sum_new_name)
+            move(old_source, new_source)
+
+        # Load the summary and statepoint files
+        summary_file = openmc.Summary(str(sum_new_name))
+        statepoint_file = openmc.StatePoint(str(sp_new_name), False)
+        statepoint_file.link_with_summary(summary_file)
+
+        # Load mgxs library
+        for mgxs in self.states[time_point].mgxs_lib.values():
+            mgxs.load_from_statepoint(statepoint_file)
+
+        self.states[time_point]._load_mgxs()
+
+    def _create_state(self, time_point):
+        """Modify state as needed to reflect current values in solver.
+
+        Parameters
+        ----------
+        time_point : str
+            Refers to the present time step of the simulation.
+
+        """
+
+        if time_point in ('START', 'END', 'FORWARD_OUTER', 'PREVIOUS_OUTER'):
+            state = openmc.kinetics.OuterState(self.states)
+            state.chi_delayed_by_delayed_group = self.chi_delayed_by_delayed_group
+            state.chi_delayed_by_mesh = self.chi_delayed_by_mesh
+            if time_point in ('FORWARD_OUTER', 'PREVIOUS_OUTER'):
+                state.method = self.method
+        else:
+            state = openmc.kinetics.InnerState(self.states)
+            state.method = self.method
+
+        state.shape_mesh = self.shape_mesh
+        state.amplitude_mesh = self.amplitude_mesh
+        state.unity_mesh = self.unity_mesh
+        state.tally_mesh = self.tally_mesh
+        state.energy_groups = self.energy_groups
+        state.fine_groups = self.fine_groups
+        state.one_group = self.one_group
+        state.num_delayed_groups = self.num_delayed_groups
+        state.use_agd = self.use_agd
+        state.use_pcmfd = self.use_pcmfd
+        state.time_point = time_point
+        state.clock = self.clock
+        state.k_crit = self.k_crit
+        state.core_volume = self.core_volume
+        state.log_file = self.log_file
```

Comment:
Earlier comment remains -- why not have `State` keep a reference to the `Solver` object and retrieve attributes from it as needed?

Reply:
This one confuses me. These definitions of `OuterState()` and `InnerState()` are identical to how we define `Solver()` in the input file and assign all of its variables. 

We are creating many `states`, each at a different `time_point`. I see that almost all of these variables don't change between time_points (mesh definitions and group boundaries, etc don't change throughout the calculation), so are you suggesting we avoid re-defining all the same variables each time a new `state` is created?

Reply:
I'm suggesting that instead of adding all these attributes to the `State` object, just keep a single attribute that refers to the `Solver` instance, i.e.,
```Python
state.solver = self
```
Then if the state object needs to access, e.g., the amplitude mesh, it can get it via `state.solver.amplitude_mesh`. Does that make sense?

Reply:
Thanks for helping me understand! I did that, and I think it works really nicely.